contracts/WooVaultManager.sol

```diff
@@ -0,0 +1,162 @@
+
+    EnumerableSet.AddressSet private vaultSet;
+
+    constructor(address newQuoteToken, address newRewardToken) public {
+        require(newQuoteToken != address(0), 'WooVaultManager: INVALID_QUOTE');
+        require(newRewardToken != address(0), 'WooVaultManager: INVALID_RAWARD_TOKEN');
+        initOwner(msg.sender);
+        quoteToken = newQuoteToken;
+        rewardToken = newRewardToken;
+    }
+
+    function allVaults() external view override returns (address[] memory) {
+        address[] memory vaults = new address[](vaultSet.length());
+        for (uint256 i = 0; i < vaultSet.length(); ++i) {
+            vaults[i] = vaultSet.at(i);
+        }
+        return vaults;
+    }
+
+    function addReward(uint256 amount) external override {
+        if (amount == 0) {
+            return;
+        }
+
+        uint256 balanceBefore = IERC20(quoteToken).balanceOf(address(this));
+        TransferHelper.safeTransferFrom(quoteToken, msg.sender, address(this), amount);
+        uint256 balanceAfter = IERC20(quoteToken).balanceOf(address(this));
+        require(balanceAfter.sub(balanceBefore) >= amount, 'WooVaultManager: amount INSUFF');
+    }
+
+    function pendingReward(address vaultAddr) external view override returns (uint256) {
+        require(vaultAddr != address(0), 'WooVaultManager: vaultAddr_ZERO_ADDR');
+        uint256 totalReward = IERC20(quoteToken).balanceOf(address(this));
+        return totalReward.mul(vaultWeight[vaultAddr]).div(totalWeight);
+    }
+
+    function pendingAllReward() external view override returns (uint256) {
+        return IERC20(quoteToken).balanceOf(address(this));
+    }
+
+    // ----------- Admin Functions ------------- //
+
+    function setVaultWeight(address vaultAddr, uint256 weight) external override onlyOwner {
+        require(vaultAddr != address(0), 'WooVaultManager: vaultAddr_ZERO_ADDR');
+
+        // NOTE: First clear all the pending reward if > 100u to keep the things fair
+        if (IERC20(quoteToken).balanceOf(address(this)) >= 1e20) {
+            distributeAllReward();
+        }
+
+        uint256 prevWeight = vaultWeight[vaultAddr];
+        vaultWeight[vaultAddr] = weight;
+        totalWeight = totalWeight.add(weight).sub(prevWeight);
+
+        if (weight == 0) {
+            vaultSet.remove(vaultAddr);
+        } else {
+            vaultSet.add(vaultAddr);
+        }
+
+        emit VaultWeightUpdated(vaultAddr, weight);
+    }
+
+    function distributeAllReward() public override onlyOwner {
+        uint256 totalRewardInQuote = IERC20(quoteToken).balanceOf(address(this));
+        if (totalRewardInQuote == 0 || totalWeight == 0) {
+            return;
+        }
+
+        uint256 balanceBefore = IERC20(rewardToken).balanceOf(address(this));
+        TransferHelper.safeApprove(quoteToken, address(wooPP), totalRewardInQuote);
+        uint256 wooAmount = IWooPP(wooPP).sellQuote(rewardToken, totalRewardInQuote, 0, address(this), address(0));
```

Comment:
recommend to initialize wooPP in `constructor`, otherwise wooPP will be zero address if not executed setWooPP before. Beside, sandwich attack maybe?

Reply:
hmmm, WooPP actually depends on FeeManager , and which depends on this vault manager. Cannot be cycle-depending back.